rplugin/python3/deoplete/sources/tag.py

```diff
@@ -24,31 +24,46 @@
 #=============================================================================
 
 import re
+from os.path import getmtime
+from collections import namedtuple
 from .base import Base
 
+TagsCacheItem = namedtuple("TagsCacheItem", "mtime tags")
+
 class Source(Base):
     def __init__(self):
         Base.__init__(self)
 
         self.mark = '[T]'
 
+        self.cache = {}
+
     def gather_candidates(self, vim, context):
         candidates = []
-
         for tags_file in vim.eval(
                 'map(tagfiles(), "fnamemodify(v:val, \\":p\\")")'):
-            with open(tags_file, 'r', errors='replace') as f:
-                candidates += parse_tags(f)
-        return [{ 'word': x } for x in list(set(candidates))]
+            mtime = getmtime(tags_file)
+            if tags_file not in self.cache or \
+                    self.cache[tags_file].mtime != mtime:
+                with open(tags_file, 'r', errors='replace') as f:
+                    candidates += parse_tags(f)
+                self.cache[tags_file] = TagsCacheItem(mtime, candidates)
+            else:
+                candidates += self.cache[tags_file].tags
+        return [{ 'word': x } for x in candidates]
 
 def debug(vim, msg):
     vim.command('echomsg string("' + str(msg) + '")')
 
 def parse_tags(f):
-    p = re.compile('^[a-zA-Z_]\w*(?=\t)')
+    p = re.compile('[a-zA-Z_]\w*(?=\t)')
```

Comment:
Understood. This is definitely cleaner, though I keep always forgetting about the fact (in most languages including VimL and lua “match” means “match anywhere”).